eras/conway/impl/test/Test/Cardano/Ledger/Conway/CommitteeRatifySpec.hs

```diff
@@ -0,0 +1,330 @@
+
+data TestData era = TestData
+  { members :: Map (Credential 'ColdCommitteeRole (EraCrypto era)) EpochNo
+  , votes :: Votes era
+  , committeeState :: CommitteeState era
+  }
+  deriving (Show)
+
+data Votes era = Votes
+  { votedYes :: [Credential 'HotCommitteeRole (EraCrypto era)]
+  , votedNo :: [Credential 'HotCommitteeRole (EraCrypto era)]
+  , votedAbstain :: [Credential 'HotCommitteeRole (EraCrypto era)]
+  , notVoted :: [Credential 'HotCommitteeRole (EraCrypto era)]
+  }
+  deriving (Show)
+
+genTestData ::
+  forall era.
+  Era era =>
+  Ratios ->
+  Gen (TestData era)
+genTestData ratios = do
+  coldCreds <- genNonEmptyColdCreds @era
+  committeeState@(CommitteeState {csCommitteeCreds}) <- genNonResignedCommitteeState @era coldCreds
+  let hotCreds = catMaybes $ Map.elems csCommitteeCreds
+      votes = distributeVotes @era ratios hotCreds
+      members = Map.fromList $ zip coldCreds (repeat (EpochNo 10))
+  pure $ TestData members votes committeeState
+
+resignMembers ::
+  [Credential 'HotCommitteeRole (EraCrypto era)] ->
+  TestData era ->
+  TestData era
+resignMembers hotCreds TestData {members, votes, committeeState} =
+  let
+    pairs = Map.toList (csCommitteeCreds committeeState)
+    updatedState =
+      [ (ck, updated)
+      | (ck, Just hk) <- pairs
+      , let updated = if hk `elem` hotCreds then Nothing else Just hk
+      ]
+   in
+    TestData members votes (committeeState {csCommitteeCreds = Map.fromList updatedState})
+
+expireMembers ::
+  [Credential 'HotCommitteeRole (EraCrypto era)] ->
+  EpochNo ->
+  TestData era ->
+  TestData era
+expireMembers hotCreds newEpochNo TestData {members, votes, committeeState} =
+  let
+    pairs = Map.toList members
+    CommitteeState {csCommitteeCreds} = committeeState
+    updatedMembers =
+      [ (ck, if expire then newEpochNo else epoch)
+      | (ck, epoch) <- pairs
+      , let expire = (maybeToList . join) (Map.lookup ck csCommitteeCreds) `List.isInfixOf` hotCreds
+      ]
+   in
+    TestData (Map.fromList updatedMembers) votes committeeState
+
+totalVotes :: Votes era -> Map (Credential 'HotCommitteeRole (EraCrypto era)) Vote
+totalVotes Votes {votedYes, votedNo, votedAbstain} =
+  Map.union
+    (Map.fromList [(cred, VoteYes) | cred <- votedYes])
+    ( Map.union
+        (Map.fromList [(cred, VoteNo) | cred <- votedNo])
+        (Map.fromList [(cred, Abstain) | cred <- votedAbstain])
+    )
+
+genNonEmptyColdCreds ::
+  forall era.
+  Era era =>
+  Gen [Credential 'ColdCommitteeRole (EraCrypto era)]
+genNonEmptyColdCreds =
+  listOf1 (arbitrary @(Credential 'ColdCommitteeRole (EraCrypto era)))
+
+genNonResignedCommitteeState ::
+  forall era.
+  Era era =>
+  [Credential 'ColdCommitteeRole (EraCrypto era)] ->
+  Gen (CommitteeState era)
+genNonResignedCommitteeState coldCreds = do
+  hotCreds <- vectorOf (length coldCreds) (arbitrary @(Credential 'HotCommitteeRole (EraCrypto era)))
+  pure $ CommitteeState $ Map.fromList $ zip coldCreds (Just <$> hotCreds)
```

Comment:
Hm this is interesting... The tests are failing if the committeeState contains duplicate values, so what you said: different cold keys mapping to the same hot key. 
The reason is that the votes are indexed by the hot key, so now instead of two votes, we'll have only 1..
I'm not sure yet if it's because of how the test data is generated - I'm sure that's at least one problem, but the the votes is still a `Map (Credential 'HotCommitteeRole (EraCrypto era)) Vote` .  
Is that what we expect, just one vote , if two cold keys map to the same hot key?

Reply:
For now I pushed a version that doesn't duplicate the hot keys, will change it back to your suggestion after figuring out the relationship of duplicated hot keys with the voting

Reply:
The only two ways it can work is one vote can be counted either once or at least once, when the same hot credential maps to many cold credentials. If I understand the spec correctly, and as I would implement it too, I would count it as many times as there are active cold credentials that match that hot credential.

So, I would suggest you try and make sure it works that way and test are passing with this too, but we should certainly discuss it on Monday meeting and see what @WhatisRT thinks about it, since I am sure he thought about this scenario.

Reply:
I adjusted the tests to support duplicate hotkeys. 
It presented some surprising challenges in terms of generating data, but it was fun to implement and I took the opportunity to also remove the duplication between generating expired and resigned members,. 
All in all I think it was a great suggestion and worth to have this case tested.  Thank you !